### PR TITLE
feishu: init at 5.9.18

### PR DESCRIPTION
**Pull request by BillHuang2001**
###### Description of changes

Add a new package Feishu, an all-in-one collaboration suite. With features like Chat, Calendar, Meetings, Docs, Mail and Workplace all in one collective app.

Homepage: [https://www.feishu.cn/en/](https://www.feishu.cn/en/)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/feishu/default.nix [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    src = fetchurl {
+      url = "https://sf3-cn.feishucdn.com/obj/ee-appcenter/${packageHash}/Feishu-linux_x64-${version}.deb";
+      sha256 = "ffb29b5a184b025d4e4ea468a8f684a7067ab5bfd45867abc370e59be63892c7";
+    };
+
+    buildInputs = [
+      autoPatchelfHook
+      makeShellWrapper # for wrapProgram
+      dpkg # for unpacking deb
+      gtk3 # for GSETTINGS_SCHEMAS_PATH
```
Comment (AndersonTorres) [inline pkgs/applications/networking/instant-messengers/feishu/default.nix:76]: Useless comments.

File: pkgs/applications/networking/instant-messengers/feishu/default.nix [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    meta = with lib; {
+      description = "An all-in-one collaboration suite";
+      homepage = "https://www.feishu.cn/en/";
+      downloadPage = "https://www.feishu.cn/en/#en_home_download_block";
+      license = licenses.unfree;
+      platforms = [ "x86_64-linux" ];
+    };
+  };
+in
+linux-feishu
```
Comment (AndersonTorres) [inline pkgs/applications/networking/instant-messengers/feishu/default.nix:175]: This is an uncommon (and useless) design pattern.

```suggestion
```

File: pkgs/applications/networking/instant-messengers/feishu/default.nix [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+, wrapGAppsHook
+, xdg-utils
+}:
+
+let
+  version = "5.9.18";
+  pname = "feishu";
+  packageHash = "5db94058d7ad"; # A hash value used in the download url
+
+  linux-feishu = stdenv.mkDerivation rec {
```
Comment (AndersonTorres) [inline pkgs/applications/networking/instant-messengers/feishu/default.nix:64]:
```suggestion
stdenv.mkDerivation rec {
  version = "5.9.18";
  pname = "feishu";
  packageHash = "5db94058d7ad"; # A hash value used in the download url
```

File: pkgs/applications/networking/instant-messengers/feishu/default.nix [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    inherit version pname;
+    system = "x86_64-linux";
```
Comment (AndersonTorres) [inline pkgs/applications/networking/instant-messengers/feishu/default.nix:66]:
```suggestion
```

Implicit.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by BillHuang2001** [comment]
Simplified the code as suggested, and this commit is squashed into the previous one.

File: pkgs/applications/networking/instant-messengers/feishu/default.nix [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+  src = fetchurl {
+    url = "https://sf3-cn.feishucdn.com/obj/ee-appcenter/${packageHash}/Feishu-linux_x64-${version}.deb";
+    sha256 = "ffb29b5a184b025d4e4ea468a8f684a7067ab5bfd45867abc370e59be63892c7";
+  };
+
+  buildInputs = [
+    autoPatchelfHook
+    makeShellWrapper
+    dpkg
```
Comment (NickCao) [inline pkgs/applications/networking/instant-messengers/feishu/default.nix:72]: These build-time dependencies should be in nativeBuildInputs.

File: pkgs/applications/networking/instant-messengers/feishu/default.nix [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    pixman
+    stdenv.cc.cc
+    systemd
+    wayland
+    xdg-utils
+  ];
+
+  dontUnpack = true;
+  installPhase = ''
+    # This deb file contains a setuid binary,
```
Comment (NickCao) [inline pkgs/applications/networking/instant-messengers/feishu/default.nix:143]: FYI, setuid binaries are stripped out of their suid bit when moved into the nix store, so does it still work without suid?

Comment (BillHuang2001) [reply pkgs/applications/networking/instant-messengers/feishu/default.nix:143]: It does work, though i don't know why there is setuid in the original file, i guess it's Electron's problem.
This workaround is copied from [slack](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/slack/default.nix)

File: pkgs/applications/networking/instant-messengers/feishu/default.nix (hunk 5 above)
Comment (BillHuang2001) [reply pkgs/applications/networking/instant-messengers/feishu/default.nix:72]: They have been moved to nativeBuildInputs now.

**Comment by gou4shi1** [comment]
Thanks for your greak work!
But I found several issues:
* My desktop is using nvidia driver, feishu will keep black screen and the logs is
```
[13010:13010:0709/134046.809136:ERROR:gpu_init.cc(441)] Passthrough is not supported, GL is angle
```
My workaround is
```
wrapProgram $out/opt/bytedance/feishu/feishu \
...
--add-flags --use-gl=desktop

wrapProgram $out/opt/bytedance/feishu/vulcan/vulcan \
...
--add-flags --use-gl=desktop
```
* Links opened from feishu can't get auth from feishu in vulcan, the logs is
```
[12992:12992:0709/134131.869338:ERROR:launch_browser_utils.cc(95)] [vulcan_lite] ebus connection not valid
[12992:12992:0709/134131.869445:ERROR:launch_browser_utils.cc(168)] [vulcan_lite] open vulcan through ipc failed
[13181:13181:0709/134132.164017:ERROR:lark_user_service.cc(534)] LarkUserService:: get lark auth info by ebus failed
```

**Comment by BillHuang2001** [comment]
I tested the flag, and with the `--use-gl=desktop` flag, it still works fine on my computer.
Still I think there might be better solution, considering `google-chrome` and some other electron apps in nixpkg don't need such flag. They were some similar issues with `google-chrome`, but it seems they're currently fixed.

**Comment by BillHuang2001** [comment]
See the new PR: https://github.com/NixOS/nixpkgs/pull/180904
Please help out if you can.
Thanks.

**Comment by gou4shi1** [comment]
Thanks for you new PR, but it doesn't help me.
```
[0711/125330.160655:INFO:log_utils.cc(286)] lark_slardar_new_client
[69630:69630:0711/125330.231321:ERROR:gpu_init.cc(441)] Passthrough is not supported, GL is angle
[69606:69624:0711/125330.785776:ERROR:resource_manager.cc(401)] [dynamic_resource] Reset data source, won't load anything, channel_name: ka_xxx, backup_dir_name: 
[69606:69606:0711/125330.831676:ERROR:lark_webview.cc(252)] OnFrameStatusChanged: unknown status!
[69606:69606:0711/125354.024374:ERROR:launch_browser_utils.cc(95)] [vulcan_lite] ebus connection not valid
[69606:69606:0711/125354.024432:ERROR:launch_browser_utils.cc(168)] [vulcan_lite] open vulcan through ipc failed
[69805:69805:0711/125354.237301:ERROR:gpu_init.cc(441)] Passthrough is not supported, GL is angle
[69767:69767:0711/125354.314528:ERROR:lark_user_service.cc(534)] LarkUserService:: get lark auth info by ebus failed
[69767:69767:0711/125354.614531:ERROR:vulcan_cloud_policy_provider.cc(170)] policy error:401
[69767:69767:0711/125357.686861:ERROR:vulcan_cloud_policy_provider.cc(170)] policy error:401
[69805:69805:0711/125359.891030:ERROR:shared_context_state.cc(73)] Skia shader compilation error
------------------------

Errors:
Program binary could not be loaded. Binary is not compatible with current driver/hardware combination. Driver build date May 27 2022. Please check build information of source that generated the binary.
[69805:69805:0711/125359.905171:ERROR:shared_context_state.cc(73)] Skia shader compilation error
------------------------
```

**Comment by BillHuang2001** [comment]
Is it the black screen problem or the auth problem?
I am currently having no solution to the auth problem, feishu is using the embedded browser and other binaries to deal with links, it's really hard to debug, since I can't execute these binaries directly.
Also, I don't have nvidia cards now, so I don't know what's wrong with nvidia's firmware. From my past experience, nvidia is really hard to work with, especially under wayland.

**Comment by gou4shi1** [comment]
> Is it the black screen problem or the auth problem?

Both.

> especially under wayland.

I'm not using wayland.

**Comment by BillHuang2001** [comment]
For the first problem, you can have to pass in additional command line arguments, for example `--use-gl=desktop` as a workaround. The new PR didn't include this flag by default, but make it easy for you to override it.

I noticed that you also wrapped `vulcan` with `--use-gl=desktop` flag, is it necessary? I only added the commandLineArgs to the `feishu` binary in the new PR, not sure if it is enough.

**Comment by gou4shi1** [comment]
> I noticed that you also wrapped `vulcan` with `--use-gl=desktop` flag, is it necessary?

Yes, otherwise, the `vulcan` opened by `feishu` will still keep black screen.

**Comment by BillHuang2001** [comment]
I am not sure about what does `vulcan` actually do, do `vulcan` and `feishu` always share the same flag, or they might need different flags?

**Comment by gou4shi1** [comment]
> I am not sure about what does `vulcan` actually do,

You can open an external link in feishu. It will open a vulcan window with that link opened inside.

**Comment by BillHuang2001** [comment]
The PR is updated, now the `vulcan` and `feishu` are wrapped in the same way. It is also possible to directly execute `vulcan`, though usually it's not needed to be executed directly.

**Comment by gou4shi1** [comment]
You have done a great job for us!

**Comment by gou4shi1** [comment]
FYI, I found that we can just disable the builtin-vulcan in the feishu setting panel :)

**Pull request merged**